File: lib/py/uapi/internal/binary/DecodeKeys.py
from typing import Any, Dict, List, Union

from uapi.internal.binary.BinaryEncoding import BinaryEncoding
from uapi.internal.binary.BinaryEncodingMissing import BinaryEncodingMissing


def decode_keys(given: Any, binary_encoder: 'BinaryEncoding') -> Any:
    if isinstance(given, dict):
        new_dict: Dict[str, Any] = {}

        for key, value in given.items():
            if isinstance(key, str):
                new_key = key
            else:
                new_key = binary_encoder.decode_map.get(key)

                if new_key is None:
                    raise BinaryEncodingMissing(key)

            encoded_value = decode_keys(value, binary_encoder)
            new_dict[new_key] = encoded_value

        return new_dict
    elif isinstance(given, list):
        return [decode_keys(item, binary_encoder) for item in given]
    else:
        return given

File: lib/ts/uapi/internal/binary/DecodeKeys.ts
import { BinaryEncoding } from 'uapi.internal.binary.BinaryEncoding';
import { BinaryEncodingMissing } from 'uapi.internal.binary.BinaryEncodingMissing';

export function decodeKeys(given: any, binaryEncoder: BinaryEncoding): any {
    if (typeof given === 'object' && given !== null) {
        const newDict: { [key: string]: any } = {};

        for (const [key, value] of Object.entries(given)) {
            let newKey: string;

            if (typeof key === 'string') {
                newKey = key;
            } else {
                const possibleNewKey = binaryEncoder.decodeMap.get(key);

                if (possibleNewKey === undefined) {
                    throw new BinaryEncodingMissing(key);
                }

                newKey = possibleNewKey;
            }

            const encodedValue = decodeKeys(value, binaryEncoder);
            newDict[newKey] = encodedValue;
        }

        return newDict;
    } else if (Array.isArray(given)) {
        return given.map((item) => decodeKeys(item, binaryEncoder));
    } else {
        return given;
    }
}
